### PR TITLE
[macOS] Add additional heuristics for context menu QR code detection

### PR DESCRIPTION
**Pull request by pxlcoder**
#### 03b876ebc5479fe38e607b80d1f09a43bccd2a8c
<pre>
[macOS] Add additional heuristics for context menu QR code detection
<a href="https://bugs.webkit.org/show_bug.cgi?id=252972">https://bugs.webkit.org/show_bug.cgi?id=252972</a>
rdar://105953041

Reviewed by Wenson Hsieh.

260897@main added initial support for QR code detection when showing context
menus. However, websites have markup-driven ways of constructing QR codes that
do not involve images. In these cases, the `HitTestResult` obtained prior to
presenting a context menu does not contain an image, and the UI process has
nothing to analyze.

This patch ports Safari heuristics to detect markup-drive QR codes into WebKit.
Specifically, &lt;svg&gt;, &lt;canvas&gt;, and &lt;table&gt; QR codes are now supported, with
additional logic to account for zooming, transforms, and partially obscured
content.

* Source/WebCore/page/ContextMenuContext.cpp:
(WebCore::ContextMenuContext::ContextMenuContext):
* Source/WebCore/page/ContextMenuContext.h:
(WebCore::ContextMenuContext::setPotentialQRCodeNodeSnapshotImage):
(WebCore::ContextMenuContext::potentialQRCodeNodeSnapshotImage const):
(WebCore::ContextMenuContext::setPotentialQRCodeViewportSnapshotImage):
(WebCore::ContextMenuContext::potentialQRCodeViewportSnapshotImage const):
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::prepareContextForQRCode):

Add a node and viewport snapshot of the element that may be a QR code, to be
sent to the UI process for analysis. These snapshots are only taken for certain
elements that are likely to contain a QR code, and if the `HitTestResult` is
missing a link and image.

Node snapshotting is unreliable, since it does not take transforms into account.
Consequently, a viewport-level snapshot is also taken. However, a viewport-level
snaphot is insufficient for partially obscured elements. Hence, two snapshots
are taken for analysis.

This approach mirrors Safari&apos;s implementation and limits the frequency of spurious
analysis. Note that taking the viewport snapshot only if QR code detection fails
for the node snapshot was considered, but this would require another IPC
roundtrip and would complicate the context menu presentation logic. The approach
used in this patch was chosen since this logic does not run frequently.

(WebCore::ContextMenuController::maybeCreateContextMenu):
* Source/WebKit/Shared/ContextMenuContextData.cpp:

Encode and decode the snapshot images as ShareableBitmaps for use in the
UI process.

(WebKit::ContextMenuContextData::ContextMenuContextData):

Drive-by cleanup. `RefPtr` does not need to be explicitly initialized.

(WebKit::ContextMenuContextData::setImage):

Drive-by cleanup to align with modern idiom used below.

(WebKit::ContextMenuContextData::setPotentialQRCodeNodeSnapshotImage):
(WebKit::ContextMenuContextData::setPotentialQRCodeViewportSnapshotImage):
(WebKit::ContextMenuContextData::encode const):
(WebKit::ContextMenuContextData::decode):
* Source/WebKit/Shared/ContextMenuContextData.h:
(WebKit::ContextMenuContextData::potentialQRCodeNodeSnapshotImage const):
(WebKit::ContextMenuContextData::potentialQRCodeViewportSnapshotImage const):
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::showAfterPostProcessingContextData):

Analyze the snapshot images, if the `WebHitTestResultData` is missing an image
and a link.

* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::qrCodeSVGString):
(TestWebKitAPI::TEST):

Add tests to exercise the various heuristics.

* Tools/TestWebKitAPI/cocoa/TestUIDelegate.h:

Add helper methods to wait for context menu presentation and extract
element information.

* Tools/TestWebKitAPI/cocoa/TestUIDelegate.mm:
(-[TestUIDelegate waitForContextMenu]):
(-[WKWebView _test_waitForContextMenu]):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView rightClickAtPoint:]):

Implement a helper method to perform a right click.

Canonical link: <a href="https://commits.webkit.org/261019@main">https://commits.webkit.org/261019@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/0828dbefee35a039ad56f4b0d9736ae5cf99dce0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110182 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19280 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42842 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1604 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119179 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114133 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20740 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10469 "Built successfully") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102437 "Reverted pull request changes (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115928 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15444 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98651 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/102437 "Reverted pull request changes (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97405 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30311 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/102437 "Reverted pull request changes (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11971 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31648 "Found 1 new test failure: media/video-playback-quality.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12583 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8616 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17948 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51265 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7634 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14390 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebKit/Shared/ContextMenuContextData.cpp [hunk 1]
```diff
@@ -90,6 +94,28 @@ void ContextMenuContextData::setImage(WebCore::Image* image)
...
+    graphicsContext->drawImage(*image, IntPoint());
+}
+
+void ContextMenuContextData::setPotentialQRCodeViewportSnapshotImage(WebCore::Image* image)
+{
+    m_potentialQRCodeViewportSnapshotImage = ShareableBitmap::create(IntSize(image->size()), { });
+    auto graphicsContext = m_potentialQRCodeViewportSnapshotImage->createGraphicsContext();
+    if (!graphicsContext)
+        return;
+    graphicsContext->drawImage(*image, IntPoint());
```
Comment (whsieh) [inline Source/WebKit/Shared/ContextMenuContextData.cpp]: Nit - ever so slightly cleaner:

```suggestion
    if (auto graphicsContext = m_potentialQRCodeViewportSnapshotImage->createGraphicsContext())
        graphicsContext->drawImage(*image, { });
```

File: Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm [hunk 2]
```diff
@@ -377,6 +382,149 @@ static void swizzledOrderFrontColorPanel(id, SEL, id)
...
+    [configuration _setContextMenuQRCodeDetectionEnabled:YES];
+
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 600, 600) configuration:configuration.get()]);
+    [webView setPageZoom:1.5];
+    [webView setUIDelegate:delegate.get()];
+    [webView synchronouslyLoadHTMLString:qrCodeSVGString()];
+    [webView mouseDownAtPoint:NSMakePoint(120, 300) simulatePressure:NO withFlags:0 eventType:NSEventTypeRightMouseDown];
+    [webView mouseUpAtPoint:NSMakePoint(120, 300) withFlags:0 eventType:NSEventTypeRightMouseUp];
+    Util::run(&gotProposedMenu);
+}
```
Comment (rmondello) [inline Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:461]: Can we extend some or all of these tests to right click outside of the element in question to ensure that no QR code is detected?

Comment (pxlcoder) [reply Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:461]: Sure.

File: Source/WebKit/Shared/ContextMenuContextData.cpp [hunk 3]
```diff
@@ -90,6 +94,28 @@ void ContextMenuContextData::setImage(WebCore::Image* image)
...
+    m_potentialQRCodeNodeSnapshotImage = ShareableBitmap::create(IntSize(image->size()), { });
+    auto graphicsContext = m_potentialQRCodeNodeSnapshotImage->createGraphicsContext();
+    if (!graphicsContext)
+        return;
+    graphicsContext->drawImage(*image, IntPoint());
+}
+
+void ContextMenuContextData::setPotentialQRCodeViewportSnapshotImage(WebCore::Image* image)
+{
+    m_potentialQRCodeViewportSnapshotImage = ShareableBitmap::create(IntSize(image->size()), { });
```
Comment (rr-codes) [inline Source/WebKit/Shared/ContextMenuContextData.cpp]: what if `image` is `nullptr`?

File: Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm [hunk 4]
```diff
@@ -512,6 +512,27 @@ static void getStandardShareMenuItem(NSArray *items, void (^completionHandler)(N
 
         return true;
     }
+
+    if (auto potentialQRCodeNodeSnapshotImage = m_context.potentialQRCodeNodeSnapshotImage()) {
+        auto image = potentialQRCodeNodeSnapshotImage->makeCGImage();
+        requestPayloadForQRCode(image.get(), [this, protectedThis = Ref { *this }](NSString *result) mutable {
+            if (!result.length) {
```
Comment (rr-codes) [inline Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm]: can this maybe be extracted into it's own method? the chain of callbacks / if statements here is confusing

File: Source/WebKit/Shared/ContextMenuContextData.cpp (hunk 3 above)
Comment (pxlcoder) [reply Source/WebKit/Shared/ContextMenuContextData.cpp]: This is a private method that has a null check at the call site. I can add an `ASSERT` though.

File: Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm (hunk 4 above)
Comment (pxlcoder) [reply Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm]: Will try to clean it up a bit.

File: Source/WebCore/page/ContextMenuContext.cpp [hunk 5]
```diff
@@ -45,6 +45,10 @@ ContextMenuContext::ContextMenuContext(Type type, const HitTestResult& hitTestRe
 #if ENABLE(SERVICE_CONTROLS)
     , m_controlledImage(nullptr)
 #endif
+#if ENABLE(CONTEXT_MENU_QR_CODE_DETECTION)
+    , m_potentialQRCodeNodeSnapshotImage(nullptr)
+    , m_potentialQRCodeViewportSnapshotImage(nullptr)
```
Comment (whsieh) [inline Source/WebCore/page/ContextMenuContext.cpp]: No need to explicitly initialize these to `nullptr`, since they're `RefPtr`.

(We could maybe even drive-by-fix the `m_controlledImage` above).

```suggestion
```

File: Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm [hunk 6]
```diff
@@ -512,6 +512,27 @@ static void getStandardShareMenuItem(NSArray *items, void (^completionHandler)(N
...
+                if (auto potentialQRCodeViewportSnapshotImage = m_context.potentialQRCodeViewportSnapshotImage()) {
+                    auto image = potentialQRCodeViewportSnapshotImage->makeCGImage();
+                    requestPayloadForQRCode(image.get(), [this, protectedThis = Ref { *this }](NSString *result) mutable {
```
Comment (whsieh) [inline Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm]: Nit - you can prevent a tiny bit of ref-churn here
```suggestion
                    requestPayloadForQRCode(image.get(), [this, protectedThis = WTFMove(protectedThis)](NSString *result) mutable {
```

**Review by whsieh — APPROVED** [review]

File: Source/WebKit/Shared/ContextMenuContextData.cpp (hunk 3 above)
Comment (pxlcoder) [reply Source/WebKit/Shared/ContextMenuContextData.cpp]: Actually, these can just be references.

File: Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm (hunk 2 above)
Comment (pxlcoder) [reply Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:461]: Updated all tests!

File: Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm (hunk 4 above)
Comment (pxlcoder) [reply Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm]: I felt it made sense to keep this in the same method, but I have re-written the logic with readability in mind, and reduced overall indentation by two levels.

**Comment by pxlcoder** [comment]
Thanks for all the reviews!
